tasks/setup-debian.yml

```diff
@@ -1,26 +1,37 @@
 ---
-- name: Update APT package cache
-  apt:
-    update_cache: true
-    cache_valid_time: 3600
-  when: ansible_distribution == "Ubuntu"
+- name: Setup WireGuard preference
+  copy:
+    src: etc/apt/preferences.d/limit-unstable
+    dest: /etc/apt/preferences.d/limit-unstable
+    owner: root
+    group: root
+    mode: 0644
   tags:
     - wg-install
 
-- name: Install required packages
-  package:
-    name: "{{ item }}"
+- name: Add WireGuard key
+  apt_key:
+    keyserver: "keyserver.ubuntu.com"
+    id: "8B48AD6246925553"
     state: present
-  with_items:
-    - software-properties-common
-    - linux-headers-{{ ansible_kernel }}
   tags:
     - wg-install
 
 - name: Add WireGuard repository
   apt_repository:
-    repo: "ppa:wireguard/wireguard"
+    repo: "deb http://deb.debian.org/debian/ unstable main"
     state: present
     update_cache: yes
   tags:
     - wg-install
+
+- name: Get architecture
+  shell: dpkg --print-architecture
+  register: dpkg_arch
+  changed_when: "0 != 0"
+
+- name: Install kernel headers to compile wireguard with DKMS
+  apt:
+    name:
+      - "linux-headers-{{ dpkg_arch.stdout }}"
```

Comment:
There is no good way to do this.

Installign `linux-headers-{{ ansible_kernel }}` would add an explicitly installed package that depends on the currently installed kernel version, which then may not be autoremoved.
Going linux-headers-arch route is not perfect either since the module will not load if the installed kernel is more modern than the running kernel (i.e. scaleways with defaults)

Reply:
Indeed it is hard to solve. I have also found an ansible role that alledgedly sets up the kernel headers for scaleways.

I propose that the module should work on "normal" debian derivatives and just fails on exceptions when the installed kernel does not match the kernel that is running.

In addition the ubuntu code may need `linux-headers-generic` package. Could add that easily and I'm ok with testing the ubuntu derivation.

Reply:
I guess this needs to be done completely different. I mean you can install `linux-image-virtual`, `linux-image-generic`, `linux-image-virtual`, `linux-image-generic-hwe`, and what not :wink: Maybe we need to allow to specify the linux kernel image name as a variable. At least optional. As a default fallback the current implementation seems to be good enough for me but allowing the user to override this.

Additionally I was thinking of a cronjob/systemd.timer that periodically checks if the installed kernel and kernel-header are matching and if not installs the kernel-header. That would at least allow to compile the WireGuard dkms.

Now the question is if a new kernel-header was installed should the job also trigger a WireGuard dkms rebuild to ensure that the new kernel has the correct WireGuard module during next reboot? Because ATM that's a problem. The role installs the kernel-header matching the currently in use kernel. And 5 minutes later a new kernel will be auto-installed but that doesn't mean it will also install the kernel-header... So if you now reboot... Tada... No WireGuard. But I also don't want to use ` apt-mark hold ...` to hold the current kernel. Nobody would expect that a WireGuard role also changes your kernel handling. I'm not really sure how to solve this problem.

Reply:
I guess this is a hard problem;

For Debian is is relatively easy. I went for the approach using `dpkg` to get the architecture since I think (did not verify) that `linux-image-amd64` would track the currently installed kernel through its dependencies.

Ubuntu is harder. I had not thought about those variations.

A default per distribution would be featible using [os-dependent-variables](https://ansible-tips-and-tricks.readthedocs.io/en/latest/os-dependent-tasks/variables/).

It is easy to get the currently running kernel name (`ansible_kernel`). Perhaps that could be installed by default on Ubuntu to match the derivation? However this would open up the problem you describe and may prevent kernels from being autoremoved.

The problem with a DKMS rebuild without a kernel version installed almost gives me the feeling that DKMS is broken by design when installed in this way 😟. I have not experienced this failure mode - but my machines likely contained `linux-image-{{ derivation }}`. And I have stayed away from scaleways…

Reply:
"enabling DKMS modules to be compiled" sounds like a valid ansible module and a good dependency for this one. And for now this module is idempotent 🙂.